### PR TITLE
Desktop login flow V2

### LINKED ISSUE #14050: Desktop login flow NG
So the login flow we have now works great in most situation. But for example doesn't work that great with special SSO or 2FA methods. Take U2F, Smart cards etc.

For this I would propose an alternative flow that can be used.

Basics
======

1. user open desktop application
2. user enters url
3. user sees a window that says something like `Click here to login in your browser`
4. user clicks, browser is opened
5. user logs in and grants access (just like it does now)
  * with all the special 2FA or SSO
6. Message `Please close this window and return to the desktop client`
7. Desktop client has obtained the token and logged in

Details
=====

At step 2 above the server will query a capability of the server that tells it what flow to use

At step 3 the client will submit a POST request to `<server>/login/ng` and obtain 
  * a `polltoken` (a long 128 char token)
  * the url to open and login `<server>/login/ng/flow/<logintoken>` this login token is also random (32 chars)

The whole login now needs to be completed within 15 minutes. After that both tokens are invalid and the flow has to be started again.

The client will now regurally (every 5 seconds) do a POST to `<server>/login/ng/poll` with the `polltoken`.
Once the authentication in step 6 is successful this will return the
* server
* loginname
* apptoken

Once step 6 is completed the `logintoken` becomes invalid
Once the server has obtained the authentication details the `polltoken` also becomes invalid

Advantages
=========

* This allows a user to login to the desktop client as long as they are able to login to the webUI
* This makes sure that we do not have to implement Kerberos/Smartcart/U2F support in the desktopclient
* If the user is already logged in they only have to grant access (and not login).

CC: @MorrisJobke @nickvergessen @ChristophWurst @blizzz @schiessle

### PR DESCRIPTION
**Pull request by rullzer**
Fixes #14050 

Basics for the login flow v2

TODO:
- [x] Rebase
- [x] Tests
- [x] Background job to cleanup expired tokens (should exist max 20 minutes)
- [x] Validation of timestamp when fetching tokens
- [x] Cleanup
- [x] Show actual client name
- [x] ~~Add throttling etc~~ leave for now. Tokens are valid for 20 minutes. Good luck trying 62^128 different ones...

Open questions:
- [x] ~~Show client ip?~~ lets not for now

### CODE REVIEW

**Comment by rullzer** [comment]
Quick review would already be appreciated before I lock it all down with tests.

File: core/Controller/ClientFlowLoginV2Controller.php [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+	 * @NoCSRFRequired
+	 * @PublicPage
+	 *
+	 * TODO: rate limiting
+	 */
+	public function init(): JSONResponse {
+		// Get client user agent
+		$userAgent = $this->request->getHeader('USER_AGENT');
+
+		//TODO: catch errors
```
Comment (MorrisJobke) [inline core/Controller/ClientFlowLoginV2Controller.php]: 🤔

**Review by MorrisJobke — APPROVED** [review]
Code looks good so far 👍

File: core/Controller/ClientFlowLoginV2Controller.php (hunk 1 above)
Comment (rullzer) [reply core/Controller/ClientFlowLoginV2Controller.php]: yeah I'll clean up when I add tests ;)

**Comment by rullzer** [comment]
tests will come tomorrow most likely. Now they are nice small fuctions that should be easy to test :D

File: core/templates/loginflowv2/grant.php [hunk 2]
```diff
@@ -0,0 +1,50 @@
+<?php
+/**
+ * @copyright Copyright (c) 2017 Lukas Reschke <lukas@statuscode.ch>
```
Comment (ChristophWurst) [inline core/templates/loginflowv2/grant.php]: :thinking:

File: lib/private/Repair/NC16/AddClenupLoginFlowV2BackgroundJob.php [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+ */
+
+namespace OC\Repair\NC16;
+
+use OC\Core\BackgroundJobs\CleanupLoginFlowV2;
+use OCP\BackgroundJob\IJobList;
+use OCP\Migration\IOutput;
+use OCP\Migration\IRepairStep;
+
+class AddClenupLoginFlowV2BackgroundJob implements IRepairStep {
```
Comment (ChristophWurst) [inline lib/private/Repair/NC16/AddClenupLoginFlowV2BackgroundJob.php]: typo: Clenup -> Cleanup

**Review by ChristophWurst — APPROVED** [review]
A lot of code but very well structured and easy to read :+1:

**Comment by rullzer** [comment]
Ah to actually test:

Do a post (anonymous) to `<server>/login/v2`

This returns josn with 2 entries:
* poll, this is the poll endpoint. you can poll by posting the token to the endpoint
* login: this is the url you have to open in your browser

Now you can continue to poll throughout the testing but it will keep returning 404s until you are done

Now complete the login and grant in the browser. Once it is done the poll will return you everything needed to login.
* server
* loginName
* appPassword

Try to do a propfind with those.
Note that the poll will only return the information once.

**Comment by ChristophWurst** [comment]
:boom:
```
Set log level to debug
Turned on maintenance mode
Updating database schema
Doctrine\DBAL\Exception\DriverException: An exception occurred while executing 'CREATE TABLE oc_login_flow_v2 (id BIGINT UNSIGNED AUTO_INCREMENT NOT NULL, timestamp BIGINT UNSIGNED NOT NULL, started SMALLINT UNSIGNED DEFAULT 0 NOT NULL, poll_token VARCHAR(255) NOT NULL, login_token VARCHAR(255) NOT NULL, public_key TEXT NOT NULL, private_key TEXT NOT NULL, client_name VARCHAR(255) NOT NULL, login_name VARCHAR(255) DEFAULT NULL, server VARCHAR(255) DEFAULT NULL, app_password VARCHAR(1024) DEFAULT NULL, UNIQUE INDEX UNIQ_57715EA23986DCFA (poll_token), UNIQUE INDEX UNIQ_57715EA2594766AF (login_token), INDEX IDX_57715EA2A5D6E63E (timestamp)) DEFAULT CHARACTER SET UTF8 COLLATE utf8_bin ENGINE = InnoDB':

SQLSTATE[42000]: Syntax error or access violation: 1075 Incorrect table definition; there can be only one auto column and it must be defined as a key
Update failed
Maintenance mode is kept active
Reset log level
```

**Comment by rullzer** [comment]
Mmm let me double check that

**Comment by rullzer** [comment]
Try again :)

**Comment by ChristophWurst** [comment]
> Try again :)

Passed :rocket:

**Comment by ChristophWurst** [comment]
Tested and works as expected :+1:

**Comment by MorrisJobke** [comment]
@rullzer Where is the documentation for this?

**Comment by rullzer** [comment]
In a local branch I need to polish and push. But I wanted this in before the freeze

**Pull request merged**